src/java/org/apache/cassandra/io/tries/TrieNode.java

```diff
@@ -0,0 +1,968 @@
+        }
+
+        @Override
+        public long lesserTransition(ByteBuffer src, int position, long positionLong, int searchIndex, long defaultValue)
+        {
+            return searchIndex == 0 || searchIndex == -1 ? defaultValue : transition(src, position, positionLong, 0);
+        }
+
+        @Override
+        public int transitionByte(ByteBuffer src, int position, int childIndex)
+        {
+            return childIndex == 0 ? src.get(position + 1) & 0xFF : Integer.MAX_VALUE;
+        }
+
+        @Override
+        public int transitionRange(ByteBuffer src, int position)
+        {
+            return 1;
+        }
+
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2 + bytesPerPointer;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            int childCount = node.childCount();
+            assert childCount == 1;
+            dest.writeByte((ordinal << 4) + (payloadBits & 0x0F));
+
+            dest.writeByte(node.transition(0));
+            writeBytes(dest, -node.serializedPositionDelta(0, nodePosition));
+        }
+    };
+
+
+    static final Single SINGLE_NOPAYLOAD_4 = new SingleNoPayload4();
+    static class SingleNoPayload4 extends Single
+    {
+        // 4-bit type ordinal
+        // 4-bit target delta
+        // byte transition
+        // no payload!
+        SingleNoPayload4()
+        {
+            super(FRACTIONAL_BYTES);
+        }
+
+        @Override
+        public int payloadFlags(ByteBuffer src, int position)
+        {
+            return 0;
+        }
+
+        // Although we don't have a payload position, provide one for calculating the size of the node.
+        @Override
+        public int payloadPosition(ByteBuffer src, int position)
+        {
+            return position + 2;
+        }
+
+        @Override
+        public long transitionDelta(ByteBuffer src, int position, int childIndex)
+        {
+            return -(src.get(position) & 0xF);
+        }
+
+        @Override
+        boolean fits(long delta)
+        {
+            return 0 <= delta && delta <= 0xF;
+        }
+
+        @Override
+        public void serialize(DataOutput dest, SerializationNode<?> node, int payloadBits, long nodePosition) throws IOException
+        {
+            assert payloadBits == 0;
+            int childCount = node.childCount();
+            assert childCount == 1;
+            long pd = -node.serializedPositionDelta(0, nodePosition);
+            assert pd > 0 && pd < 0x10;
+            dest.writeByte((ordinal << 4) + (int) (pd & 0x0F));
+            dest.writeByte(node.transition(0));
+        }
+
+        @Override
+        public int sizeofNode(SerializationNode<?> node)
+        {
+            return 2;
+        }
+    };
+
+    static final Single SINGLE_NOPAYLOAD_12 = new SingleNoPayload12();
```

Comment:
nit: Should probably move these into `SingleNoPayload12` itself to avoid potential classloading deadlocks